### DISCUSSION TITLE
SPAM via OpenMage forms - The Cloudflare Turnstile solution

### DISCUSSION THREAD

**Discussion by toejam34** [OpenMage/magento-lts#3886] [Questions & Answers] [answered]
Hello all, 

I have seen there are few somewhat outdated prevent spam on contact forms. 

Does anyone have, or know of any that might use a more modern approach such as turnstile? 

Any pointers are appreciated! 

Andy

**Comment by pquerner** [comment]
The methods havent changed.

- Honeypot
- Captcha (Google or anything else)
- Secure w/ something like visitor must be firewalled from Cloudflare (i.e. not on their blocklist)
- Blocklist by mail or domain, with false positive return

**Reply by toejam34** [reply]
I am aligned with you on the strategy , but I am curious on implementation. 

A lot of the M1 plugins do not appear to work with OM now the game has moved on. 

I am curious if anyone is using items such as turnstile or recaptcha.

**Reply by pquerner** [reply]
The strategy hasnt changed. Add a captcha of your choice in the frontend, add the formfield to the POST request of your form (if you are using standard `<form>` thats done for you by the browser, if the field is within `<form>`), and in the backend, patch the controller action to first check if the captcha was solved.

For example for Google Recaptcha: https://developers.google.com/recaptcha/docs/verify

So any information you find regarding this online (for Magento 1) are still perfectly valid.
I also dont see any reason why an existing module for this should break. They should also just simply work.

**Reply by fballiano** [reply]
I don't think a lot of plugins don't with with OM, the only "problems" I've found were because of php8, but not OM

**Reply by ma4nn** [reply]
Also https://github.com/magento-hackathon/HoneySpam is working fine with OM.

**Reply by toejam34** [reply]
I think you are absolutely right here. My terminology was inaccurate.

**Reply by toejam34** [reply]
Thank you - investigating

**Reply by fballiano** [reply]
I've added HoneySpam to https://github.com/fballiano/awesome-openmage

**Reply by empiricompany** [reply]
Clean implementation of Google reCaptcha for OpenMage: https://github.com/empiricompany/reCaptcha/

**Comment by addison74** [comment]
I think I have discussed this topic in other posts. I use HoneySpam together with Fail2Ban. Around 120 IPs are captured daily on the contact form and newsletter. From time to time a few messages slip out, but we find the senders in the webserver log by date and block them with ipset. I created a command that quickly finds the IP in CLI. We can use reCaptcha, but I gave up. It seems that it is not so secured anymore, there are articles on the Internet related to this issue. Also, visitors no longer have the patience to solve puzzles or enter codes that are not legible.

**Reply by toejam34** [reply]
Noted - thank you

**Reply by addison74** [reply]
Check this post for reCaptcha

https://github.com/OpenMage/magento-lts/discussions/1649

**Comment by addison74** [comment]
**Cloudflare Turnstile** could be a great addition for OpenMage.

https://www.cloudflare.com/products/turnstile/ 

_Humans spend about **500 years each day solving CAPTCHAs** - wasting time and creating frustration. This is why Cloudflare built a better alternative to CAPTCHAs with Managed Challenge and now Turnstile._

**Reply by f1-outsourcing** [reply]
What about this part1. I am going to try this soon in production, see if it is a quick and easy solution.

https://www.nfriedly.com/techblog/2009/11/how-to-build-a-spam-free-contact-forms-without-captchas/

**Reply by pquerner** [reply]
Thats a honeypot. If you are really spammed a lot, you can do this aswell. Doesnt hurt actual people.

**Comment by fballiano** [comment, marked as answer]
I started to work on a Cloudflare Turnstile module:
https://github.com/fballiano/openmage-cloudflare-turnstile

It already works although it supports only customer-registration, contact-form and newsletter-subscribe at the time of writing this comment.

It doesn't depend of Mage_Captcha for anything so Mage_Captcha could be disabled/removed (recaptha module depends on Mage_Captcha instead).

:-)

**Reply by addison74** [reply]
I'm glad you caught the ball :)

**Reply by fballiano** [reply]
I think I've added support for all the forms already managed by mage_captcha/recaptcha (frontend/backend)

**Reply by toejam34** [reply]
Incredible. Simply incredible. 

Thank you so much @fballiano . Coffee on me.

**Reply by addison74** [reply]
I highly recommend to all testing @fballiano's extension. So far I am happy with the results. We need a feedback from a production environment. As a visitor I am happy filling in a form without solving stupid puzzles. 

We can go further to let people know how to customize the widget (light/dark mode, ...)

**Reply by addison74** [reply]
Cloudflare has provided a way that no one has thought before. Submitting a form is not a test of intelligence (mathematics), recognition of traffic lights, buses, animals, directions and especially ophthalmology.

**Reply by toejam34** [reply]
I have implemented this in production and am monitoring spam as we speak.

0 reports and been in 12h.

**Reply by addison74** [reply]
I did not use CAPTCHA before because I wanted to avoid the situation when the service provided by a third party is non-functional. I would have risked not receiving messages, orders, for a period of time.

For the beginning I will do the production testing only on the contact form and I will periodically check the webserver logs. I am using HoneySpam + Fail2Ban with great results catching the IP's daily but I will disable it, to avoid possible conflicts.

**Reply by toejam34** [reply]
~32h in and zero issues to report. 

Quite literally zero spam has come through and I have seen orders being placed.

**Reply by fballiano** [reply]
@toejam34 great to know it's working fine!

**Reply by toejam34** [reply]
Never any doubt!

**Reply by addison74** [reply]
@toejam34 - What type of widget did you use? The invisible one?

**Reply by toejam34** [reply]
I left all options as default - literally just punched in the keys and off I went. 

Protection on the admin / contact forms.

![image](https://github.com/OpenMage/magento-lts/assets/57115662/34110102-cccc-4d26-9898-07c323926b27)

**Reply by fballiano** [reply]
2 more customers tell me that with turnstile (with my module) the spam problem vanished completely

**Comment by toejam34** [comment]
@fballiano I have found one issue. 

If you try and reset your password I get a "invalid captcha" 

I have validated that its the awesome plugin that you have written by disabling the reset password works.

Steps to reproduce : 

OM20.6
PHP8.3
URL : https://www.seriouslylotus.com/
Click login OR my account
Click forgot your password , type in any email address and press submit
This results in a "incorrect CAPTCHA" error : 

![image](https://github.com/OpenMage/magento-lts/assets/57115662/1f73393d-4561-48c6-9b79-111bc7d72354)

Edit : updated to the latest version and issue has gone!

**Reply by fballiano** [reply]
exactly, in 0.4.0 i've fixed it:
https://github.com/fballiano/openmage-cloudflare-turnstile/releases/tag/0.4.0
